### PR TITLE
Result predicate function for circuitbreaker

### LINKED ISSUE #384: CircuitBreaker record a failure when a certain result is returned
`Retry` allows check the result of a call and retry a call when a configurable predicate is true.
It would be nice if the CircuitBreaker allows it as well.
`CircuitBreakerConfig` should allow to configure a result predicate.

### PR DESCRIPTION
**Pull request by lespinsideg**
Resolves #384 

- added result predicate function in execute and decorate methods.

### CODE REVIEW

File: resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreaker.java [hunk 1]
```diff
@@ -556,13 +709,32 @@ static CheckedRunnable decorateCheckedRunnable(CircuitBreaker circuitBreaker, Ch
...
+    static <T> Consumer<T> decorateConsumer(CircuitBreaker circuitBreaker, Consumer<T> consumer, Function1<Long, Boolean> resultPredicate){
         return (t) -> {
             CircuitBreakerUtils.isCallPermitted(circuitBreaker);
             long start = System.nanoTime();
             try {
                 consumer.accept(t);
                 long durationInNanos = System.nanoTime() - start;
-                circuitBreaker.onSuccess(durationInNanos);
+
+                if (resultPredicate.apply(durationInNanos)) {
```
Comment (Romeh) [inline resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreaker.java]: @lespinsideg  the result predicate take the start time any reason why ? as u can simply do the calculation inside the if/else statement blocks .

**Comment by RobWin** [comment]
Thank you for your PR.
But I'm not sure if we should add a `resultPredicate` to all methods.
In `RetryConfig` we added an option to configure a global `resultPredicate`.  But I actually prefer higher-order functions.

If someone wants to be able to have a different resultPredicate per decorated function, he could easily use a higher-order function.

```
static <T> Supplier<T> checkResult(Supplier<T> supplier, Function<T, T> exceptionMapper){
        return () -> {
            T result = supplier.get();
            return exceptionMapper.apply(result);
        };
}
```

and then something like

```
        Supplier<HttpResponse> httpResponseSupplier = checkResult(helloWorldService::returnResult, result -> {
            if (result.getStatusCode() == 400) {
                throw new ClientException();
            } else if (result.getStatusCode() == 500) {
                throw new ServerException();
            }
            return result;
        });
        //When
        HttpResponse httpResponse = circuitBreaker.executeSupplier(httpResponseSupplier);
```

**Comment by storozhukBM** [comment]
Agree with @RobWin

**Comment by storozhukBM** [comment]
Sorry, closed by mistake.

File: resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreaker.java (hunk 1 above)
Comment (lespinsideg) [reply resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreaker.java]: @Romeh Sorry but I didn't get your point. The result predicate doesn't take the start time. I thought duration could be a parameter for deciding the result.
https://github.com/resilience4j/resilience4j/issues/381

**Comment by lespinsideg** [comment]
@RobWin @Romeh @storozhukBM 
I will check `RetryConfig` and update the code soon. Thanks for the review.

**Comment by RobWin** [comment]
I really think it was a mistake to add a onResult predicate to Retry. A decorator would have been the better choice. A decorator could have been reused for CircuitBreaker as well.

**Comment by RobWin** [comment]
We added it to Retry, because we didn't want to throw an exception every time we retry. But in our CirucirBreaker we need an exception to be thrown.

**Pull request closed without merging**